src/lib/transition_router/transition_router.ml

```diff
@@ -139,6 +122,137 @@ module Make (Inputs : Inputs_intf) = struct
+                      their best tip"
+                    peer
+                in
+                Logger.warn logger ~module_:__MODULE__ ~location:__LOC__
+                  ~metadata:[("error", `String (Error.to_string_hum e))]
+                  !"%s" error_msg ;
+                let%map () =
+                  Trust_system.(
+                    record trust_system logger peer.host
+                      Actions.(Violated_protocol, Some (error_msg, [])))
+                in
+                acc
+            | Ok (`Root _, `Best_tip candidate_best_tip) ->
+                let enveloped_candidate_best_tip =
+                  Envelope.Incoming.wrap ~data:candidate_best_tip
+                    ~sender:(Envelope.Sender.Remote peer.host)
+                in
+                return
+                @@ Option.merge acc
+                     (Option.return enveloped_candidate_best_tip)
+                     ~f:(fun existing_best_tip candidate_best_tip ->
+                       Envelope.Incoming.max
+                         ~f:(fun t1 t2 ->
+                           External_transition.compare
+                             (With_hash.data @@ fst t1)
+                             (With_hash.data @@ fst t2) )
+                         existing_best_tip candidate_best_tip ) ) )
+
+  let initialize ~logger ~network ~verifier ~trust_system ~frontier
+      ~time_controller ~ledger_db ~frontier_w ~proposer_transition_reader
+      ~clear_reader ~verified_transition_writer ~transition_reader_ref
+      ~transition_writer_ref ~most_recent_valid_block_writer =
+    let%bind () =
+      let connectivity_time_upper_bound =
+        Consensus.Constants.initialization_time_in_secs
+      in
+      let high_connectivity_deferred =
+        Ivar.read (Network.high_connectivity network)
+      in
+      Deferred.any
+        [ high_connectivity_deferred
+        ; ( after (Time_ns.Span.of_sec connectivity_time_upper_bound)
+          >>| fun () ->
+          if not @@ Deferred.is_determined high_connectivity_deferred then
+            Logger.info logger ~module_:__MODULE__ ~location:__LOC__
+              ~metadata:
+                [ ("num_peers", `Int (List.length @@ Network.peers network))
+                ; ( "connectivity_time_upper_bound"
+                  , `Float connectivity_time_upper_bound ) ]
+              !"After waiting $connectivity_time_upper_bound, we only \
+                accumulate $num_peers peers. Possible reasons for that could \
+                be: \n\
+                1. You're the first couple of peers joining the testnet. \n\
+                2. The network is slow. \n\n\n\
+                Would start initialization anyway." ) ]
+    in
+    match%map download_best_tip ~logger ~network ~verifier ~trust_system with
+    | Some best_tip_enveloped ->
+        best_tip_enveloped |> Envelope.Incoming.data |> fst |> With_hash.data
+        |> Broadcast_pipe.Writer.write most_recent_valid_block_writer
```

Comment:
Tiny nit: put this inside of download_best_tip so that every time we see a better tip, we write it to the `most_recent_valid_block_writer`.